File: MacTemplet/View/NNTableView.m
//
//  NNTableView.m
//  MacTemplet
//
//  Created by Bin Shang on 2019/6/12.
//  Copyright © 2019 Bin Shang. All rights reserved.
//

#import "NNTableView.h"

@interface NNTableView ()
@property (nonatomic, strong) NSScrollView *scrollView;

@end

@implementation NNTableView

- (instancetype)init{
    self = [super init];
    if (self) {
        self.scrollView.documentView = self;
        
    }
    return self;
}

- (void)drawRect:(NSRect)dirtyRect {
    [super drawRect:dirtyRect];
    
    // Drawing code here.
}

-(NSScrollView *)scrollView{
    if (!_scrollView) {
        _scrollView = ({
            NSScrollView *view = [[NSScrollView alloc] init];
            view.backgroundColor = NSColor.redColor;
            
            view.drawsBackground = false;//不画背景（背景默认画成白色）
            view.hasHorizontalScroller = true;
            view.hasVerticalScroller = true;
            view.autohidesScrollers = YES;//自动隐藏滚动条（滚动的时候出现）
            view;
        });
    }
    return _scrollView;
}

@end

File: MacTemplet/NNBaseView/NNTableView.swift
//
//  NNTableView.swift
//  MacTemplet
//
//  Created by Bin Shang on 2020/4/1.
//  Copyright © 2020 Bin Shang. All rights reserved.
//

import Cocoa

class NNTableView: NSTableView {
    
    lazy var scrollView: NSScrollView = {
        let scrollView = NSScrollView()
        scrollView.backgroundColor = NSColor.white
        scrollView.drawsBackground = false
        scrollView.hasHorizontalScroller = true
        scrollView.hasVerticalScroller = true
        scrollView.autohidesScrollers = true
        
        return scrollView
    }()
    
    var adjustsTableColumnsWidth = true
    
    // MARK: -lifecycle    
    override init(frame frameRect: NSRect) {
        super.init(frame: frameRect)
        
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func draw(_ dirtyRect: NSRect) {
        super.draw(dirtyRect)

        // Drawing code here.
    }
    
    override func layout() {
        super.layout()
        
        if adjustsTableColumnsWidth == true {
            for column in tableColumns {
                column.width = scrollView.bounds.width/CGFloat(tableColumns.count);
            }
        }
    }
    
    // MARK: -funtions
    func setupUI() {
        scrollView.documentView = self
        
    }
    
}
